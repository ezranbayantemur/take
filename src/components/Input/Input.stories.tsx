import React from 'react';
import {Alert, FlatList, Text, View} from 'react-native';
import {Input} from '@components';
import Button from './Input';
import type {Meta, StoryObj} from '@storybook/react-native';
import type {InputProps} from './Input.types';

const InputMeta: Meta<InputProps> = {
  argTypes: {
    onChangeText: {
      action: 'on change text',
    },
  },
  args: {
    placeholder: 'Ara...',
    debounceTime: 0,
  },
  title: 'Input',
  component: Input,
};

export default InputMeta;

export const Basic: StoryObj<InputProps> = {};

export const WithDebounce: StoryObj<InputProps> = {
  args: {
    debounceTime: 500,
  },
  decorators: [
    (_, options) => {
      const [textList, setTextList] = React.useState<string[]>([]);

      function handleChangeText(text: string) {
        setTextList(list => [...list, text]);
      }

      return (
        <View>
          <Input {...options.args} onChangeText={handleChangeText} />
          <FlatList
            data={textList}
            renderItem={({item}) => <Text>Â· {item}</Text>}
          />
        </View>
      );
    },
  ],
};
