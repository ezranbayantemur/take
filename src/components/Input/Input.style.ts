import {StyleSheet} from 'react-native';
import {color, padding, font, typography} from '@style';

export default StyleSheet.create({
  container: {
    backgroundColor: color.MERCURY,
    padding: padding.MEDIUM,
    borderRadius: 5,
    margin: padding.SMALL,
  },
  input: {
    fontFamily: font.REGULAR,
    fontSize: typography.NORMAL,
  },
});
