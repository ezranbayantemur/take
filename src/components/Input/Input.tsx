import React from 'react';
import {Text, TextInput, View} from 'react-native';
import {useDebouncedCallback} from 'use-debounce';
import styles from './Input.style';
import type {InputProps} from './Input.types';

const Input = ({
  testID,
  debounceTime = 0,
  placeholder,
  errorMessage,
  onChangeText,
  ...rest
}: InputProps) => {
  const hasError = !!errorMessage;
  const styleKey = hasError ? 'error' : 'default';

  const debouncedChangeText = useDebouncedCallback(value => {
    if (onChangeText) {
      onChangeText(value);
    }
  }, debounceTime);

  const changeTextFunction =
    debounceTime !== 0 ? debouncedChangeText : onChangeText;

  return (
    <View>
      <View style={styles[styleKey].container}>
        <TextInput
          {...rest}
          style={styles[styleKey].input}
          testID={`${testID}_input`}
          placeholder={placeholder}
          onChangeText={changeTextFunction}
        />
      </View>
      {hasError && (
        <Text
          testID={`${testID}_input_error_message`}
          style={styles.error.error_text}>
          {errorMessage}
        </Text>
      )}
    </View>
  );
};

export default React.memo(Input);
