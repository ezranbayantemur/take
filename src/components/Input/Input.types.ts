import type {TextInputProps} from 'react-native';

export interface InputProps extends TextInputProps {
  debounceTime?: number;
  errorMessage?: string;
}
