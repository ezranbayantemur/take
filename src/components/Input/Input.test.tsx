import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import Input from './Input';

let wrapper: ReturnType<typeof render>;

describe('Input unit tests', () => {
  beforeEach(() => {
    wrapper = render(<Input />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render default placeholder correctly', () => {
    const placeholder = wrapper.queryByPlaceholderText('Ürün ara...');
    expect(placeholder).not.toBeNull();
  });

  it('should render given placeholder correctly', () => {
    const _wrapper = render(<Input placholder="test placeholder" />);
    const placeholder = _wrapper.queryByPlaceholderText('test placeholder');
    expect(placeholder).not.toBeNull();
  });

  it('should trigger on debounce if it is given', () => {
    const mockOnChangeText = jest.fn();
    const _wrapper = render(
      <Input
        testID="basetest"
        onChangeText={mockOnChangeText}
        debounceTime={250}
      />,
    );

    const input = _wrapper.getByTestId('basetest_input');
    fireEvent.changeText(input, 'test');

    expect(mockOnChangeText).not.toBeCalled();

    jest.advanceTimersByTime(250);

    expect(mockOnChangeText).toBeCalled();
  });
});
