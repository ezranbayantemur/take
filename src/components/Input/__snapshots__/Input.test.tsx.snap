// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input unit tests should render correctly 1`] = `
<View>
  <View
    style={
      {
        "backgroundColor": "#E1E1E1",
        "borderRadius": 5,
        "margin": 8,
        "padding": 16,
      }
    }
  >
    <TextInput
      placeholder="Ürün ara..."
      style={
        {
          "fontFamily": "Lato-Regular",
          "fontSize": 15,
        }
      }
      testID="undefined_input"
    />
  </View>
</View>
`;
